### PR TITLE
Build docs on GH Actions

### PR DESCRIPTION
**Pull request by banesullivan**
Changes the CI to build the Docs on GH Actions.

This uses the new vtk-osmesa wheel published in pyvista-wheels to simplify setting up the environment.

When the documentation is built, it's HTML output are uploaded as a build artifact so that we can easily download the built documentation to evaluate before publishing.

I added a workflow_dispatch trigger that can be used to reset the cache on-demand. This is useful if we are debugging things like transparency in a PR and need to see how the documentation looks after being completely rebuilt

To do:

- [x] Copy `README.md ` and `ads.txt`
- [x] Create CNAME records (`echo "docs.pyvista.org" >> CNAME`)
- [x] Cache examples gallery, _build directory, and source data for examples
- [x] Push example notebooks to `pyvista-examples` `main` and `develop` branches
- [x] Turn of builds for `no-ci/` branches, etc.
- [x] Archive build directory for artifact uploads to improve speed
- [x] Clear cache on releases
- [x] Verify/Test deployment jobs

### CODE REVIEW

**Comment by banesullivan** [comment]
I'll see if I can come back to this over the holidays

**Comment by MatthewFlamm** [comment]
If this works and is possible, it would be nice to publish an artifact if successfully built.  Then we can review/check the documentation right on the CI.

File: .github/workflows/docs.yml [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+      - name: Install PyVista
+        run: pip install -e .
+      - name: Install Dependencies and PyVista
+        run: |
+          sudo apt-get install python3-tk pandoc
+          pip install -r requirements_docs.txt
+      - name: PyVista Report
+        run: python -c "import pyvista;print(pyvista.Report())"
+      - uses: actions/cache@v2
+        if: ${{ (github.event.inputs.cache && github.event_name == 'workflow_dispatch') || github.event_name != 'workflow_dispatch' }}
```
Comment (banesullivan) [inline .github/workflows/docs.yml:44]: This needs to be looked at/tested. Also, how can we have this be `false` on tags?

**Comment by banesullivan** [comment]
Ready to start reviewing. The deployment/publishing jobs need to be carefully double-checked.

**Comment by akaszynski** [comment]
Two changes, one minor and one major.

@MatthewFlamm, I had to remove `reset_pyvista` as this causes the doc build to reset each time. As it is, builds are still mostly resetting, but it's at least fixing it locally for me.

Added a virtual environment cache in hopes that there was a way to avoid resetting the doc build. Turns out that's not the solution, but it does shave off our build time by a minute. Cache key is the hash of our `requirements_docs.txt`, so I think it's safe to say that the cache won't affect our build quality. In fact, it's something we might consider applying to the rest of the project, especially with dependabot.

---

That being said, I'm not able to figure out why `sphinx` thinks our source files are reset between builds with no source changes:

```
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 2979 source files that are out of date
updating environment: 0 added, 2858 changed, 0 removed
```

Locally, the culprit for a full rebuild was `reset_pyvista` as the method contains a different memory address for each run ([source](https://groups.google.com/g/sphinx-users/c/G_Z1yQrMeTo/m/NOVNSN6ICwAJ)). However, in our CI/CD, despite caching `doctree`, the documentation is still being built each time and because of our `autosummary` directives, doc build is still taking an annoying 30 minutes.

More perplexing is creating a new virtual environment, installing requirements, building, and then repeating also causes a full rebuild locally. Steps followed (with two ):

```
python -m venv <venv-name>
source <venv-name>/bin/activate
cd ~/pyvista
pip install -e .
pip install -r requirements_docs.txt
make -C doc html
```

However, just running:
```
make -C doc html
make -C doc html
```
Causes the second build to complete without a full rebuild (only 10 files are rebuilt).

@larsoner, do you have any insight into what could be causing this?

**Comment by larsoner** [comment]
> @MatthewFlamm, I had to remove reset_pyvista as this causes the doc build to reset each time. As it is, builds are still mostly resetting, but it's at least fixing it locally for me.

Instead of removing, you can refactor into a class like (see the warning [here](https://sphinx-gallery.github.io/stable/configuration.html#sorting-gallery-subsections) for example):
```
class ResetPyvista:
    def __call__(self):
        ... # do the resetting

    def __repr__(self):
        return 'ResetPyvista'

reset_pyvista = ResetPyvista()
```
and then `reset_pyvista` will have a stable `__repr__` across runs.

> building [html]: targets for 2979 source files that are out of date

For this, is it possible that the cache that you restore does not properly restore modified times by adding some `ls -al` commands before and after the doc build? IIRC this is what Sphinx uses to determine out-of-dateness of individual files. You aren't getting a warning about the config changed, so it seems like it probably is the modified times... what's funny is that you do get `[new config]` in e4b62aa (last commit before removing reset_pyvista) but it shows only "54 source files that are out of date", whereas in the latest commit you get "2979 source files that are out of date" (assuming I should be looking at the Build Documentation" run).

For debugging purposes, I would also try adding a second `        run: make -C doc html SPHINXOPTS="-w build_errors.txt -N"` step to the CI to make sure that it does in fact *not* rebuild the whole site.

**Comment by MatthewFlamm** [comment]
> Instead of removing, you can refactor into a class like (see the warning [here](https://sphinx-gallery.github.io/stable/configuration.html#sorting-gallery-subsections) for example):

I can confirm that using a callable class works locally for the examples.  It would be useful to have this warning for `reset_modules` as well in that documentation.

> For this, is it possible that the cache that you restore does not properly restore modified times by adding some ls -al commands before and after the doc build?

Something like https://github.com/MestreLion/git-tools `restore-mtime` could be used.

**Comment by banesullivan** [comment]
Deployment jobs have been tested and are working. This is ready for final review and merge

Failing Style and Docstring CI appears unrelated

With this comes: https://github.com/pyvista/pyvista-examples/pkgs/container/pyvista-examples

**Comment by akaszynski** [comment]
> Deployment jobs have been tested and are working. This is ready for final review and merge

Going to try a final go with caching our `_autosummary` files.

@MatthewFlamm's suggestion of using `restore-mtime` works for our main source documentation and python source, but as our autogenerated docs are rebuilt each time `0 added, 2802 changed, 0 removed` we need a way of both caching that *and* making sore the cached files don't include removed methods. I might not be able to fix this.

**Comment by akaszynski** [comment]
> Going to try a final go with caching our _autosummary files.

Actually, let's merge this and then try to speed it up in a separate PR.

**Review by akaszynski — APPROVED** [review]
Nice work and LGTM. I'll attempt caching autosummary in a follow-up PR.

**Pull request merged**